src/validation.cpp

```diff
@@ -975,13 +1006,89 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
     assert(std::all_of(txns.cbegin(), txns.cend(), [this](const auto& tx)
                        { return !m_pool.exists(GenTxid::Txid(tx->GetHash()));}));
 
+    // Populate with the union of all transactions' ancestors.
+    CTxMemPool::setEntries m_collective_ancestors;
+    for (const auto& ws : workspaces) {
+        for (const auto& it : ws.m_ancestors) m_collective_ancestors.insert(it);
+    }
+
     std::string err_string;
     if (!m_pool.CheckPackageLimits(txns, m_limit_ancestors, m_limit_ancestor_size, m_limit_descendants,
                                    m_limit_descendant_size, err_string)) {
+        // TODO: When replacements exist, to avoid overestimating descendant counts, deduct the
+        // to-be-replaced mempool entries when counting descendants. Note that this is not
+        // necessarily as simple as subtracting the count/size from descendant limits, because
+        // multiple transactions may conflict with the same entries, causing us to double-count them
+        // and their descendants.
```

Comment:
I mean a situation like this:

<img width="673" alt="image" src="https://user-images.githubusercontent.com/25183001/168615825-6251d4d6-97b6-45cf-ada7-91c6a1d9ce45.png">


{A, B, C} are in the mempool, and A is already at the descendant limit (101KvB). {X, Y, Z} is a package where X conflicts with B and will replace {B, C}. However, when we run descendant limit checks, it will look like A's descendants include {B, C, X, Z} and reject it. But actually A's descendants will only include {X, Z} after the replacement.